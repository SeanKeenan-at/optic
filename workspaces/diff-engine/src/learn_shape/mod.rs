use crate::shapes::JsonTrail;
use crate::HttpInteraction;
use std::collections::{HashMap, HashSet};

mod trail_values;
mod traverser;
mod visitors;

pub type FieldSet = HashSet<String>;
pub use trail_values::{for_body_descriptor as trail_values_for_body, TrailValueMap};

#[derive(Debug)]
pub struct TrailValues {
  trail: JsonTrail,
  was_string: bool,
  was_number: bool,
  was_boolean: bool,
  was_null: bool,
  was_array: bool,
  was_object: bool,
  field_set: HashSet<FieldSet>,
}

impl TrailValues {
  pub fn new(json_trail: &JsonTrail) -> Self {
    Self {
      trail: json_trail.clone(),
      was_string: false,
      was_number: false,
      was_boolean: false,
      was_null: false,
      was_array: false,
      was_object: false,
      field_set: Default::default(),
    }
  }
}
