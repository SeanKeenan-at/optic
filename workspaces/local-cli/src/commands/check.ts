import { Command } from '@oclif/command';
import { verifyTask } from '../shared/verify/verify';

export default class Check extends Command {
  static description =
    'verify that Optic can run your tasks and monitor traffic';

  static args = [
    {
      name: 'taskName',
    },
  ];

  async run() {
    const { args } = this.parse(Check);
    const { taskName } = args;
    await verifyTask(this, taskName);
    process.exit(0);
  }
}
