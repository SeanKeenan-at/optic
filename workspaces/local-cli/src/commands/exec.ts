import { Command, flags } from '@oclif/command';
import {
  LocalCliTaskFlags,
  LocalTaskSessionWrapper,
  runCommandFlags,
} from '../shared/local-cli-task-runner';
import { cleanupAndExit, loadPathsAndConfig } from '@useoptic/cli-shared';
import { isCommandOnlyTask } from '@useoptic/cli-config';

export default class Exec extends Command {
  static description =
    'run a command and receives traffic over an ingest-service';

  static flags = runCommandFlags;

  static args = [
    {
      name: 'command',
    },
  ];

  async run() {
    const { args } = this.parse(Exec);
    const { flags } = this.parse(Exec);
    const { taskName } = args;
  }
}

export function runExec(
  cli: Command,
  command: string,
  flags: LocalCliTaskFlags
) {}
