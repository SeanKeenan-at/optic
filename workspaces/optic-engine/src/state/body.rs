use crate::shapehash;
use serde::de::value;
use serde_json::map::Map as JsonMap;
use serde_json::Value as JsonValue;
use serde_urlencoded;
use std::collections::{BTreeMap, HashMap};

#[derive(PartialEq, Clone, Debug, Hash, Eq)]
pub enum BodyDescriptor {
  Object(ObjectDescriptor),
  Array(ItemsDescriptor),
  String,
  Number,
  Boolean,
  Null,
}

impl BodyDescriptor {
  pub fn empty_object() -> Self {
    Self::Object(ObjectDescriptor::from(std::iter::empty()))
  }
}

#[derive(PartialEq, Clone, Debug, Hash, Eq)]
pub struct FieldDescriptor(pub String, pub Box<BodyDescriptor>);

impl FieldDescriptor {
  pub fn new(key: String, body: BodyDescriptor) -> Self {
    Self(key, Box::new(body))
  }
}

#[derive(PartialEq, Clone, Debug, Hash, Eq)]
pub struct ObjectDescriptor {
  fields: Vec<FieldDescriptor>,
}

impl ObjectDescriptor {
  pub fn keys(&self) -> impl Iterator<Item = &String> {
    self.fields.iter().map(|FieldDescriptor(key, body)| key)
  }

  pub fn entries(self) -> impl Iterator<Item = (String, BodyDescriptor)> {
    self
      .fields
      .into_iter()
      .map(|FieldDescriptor(key, body)| (key, *body))
  }
}

impl<T> From<T> for ObjectDescriptor
where
  T: Iterator<Item = (String, BodyDescriptor)>,
{
  fn from(entries: T) -> Self {
    Self {
      fields: entries
        .map(|(key, value)| FieldDescriptor::new(key, value))
        .collect(),
    }
  }
}

#[derive(PartialEq, Clone, Debug, Hash, Eq)]
pub struct ItemsDescriptor {
  unique_items: Box<Vec<(BodyDescriptor, Vec<usize>)>>,
}

impl<T> From<T> for ItemsDescriptor
where
  T: Iterator<Item = BodyDescriptor>,
{
  fn from(all_items: T) -> Self {
    let mut unique_items = all_items
      .enumerate()
      .fold(HashMap::new(), |mut indexes_by_unique_items, (i, item)| {
        {
          let items = indexes_by_unique_items.entry(item).or_insert(vec![]);
          items.push(i);
        };

        indexes_by_unique_items
      })
      .into_iter()
      .map(|(item, indexes)| (item.clone(), indexes))
      .collect::<Vec<_>>();

    unique_items.sort_by_key(|(_, indexes)| *indexes.first().unwrap());

    Self {
      unique_items: Box::new(unique_items),
    }
  }
}

impl From<BodyDescriptor> for ItemsDescriptor {
  fn from(single_item: BodyDescriptor) -> Self {
    Self {
      unique_items: Box::new(vec![(single_item, vec![0])]),
    }
  }
}

impl ItemsDescriptor {
  // @TODO: decide if this is still interesting at all. Could be reconstructed from unique_items
  // pub fn into_all(self) -> impl Iterator<Item = BodyDescriptor> {
  //
  // }

  pub fn into_unique(self) -> impl Iterator<Item = (BodyDescriptor, Vec<usize>)> {
    self.unique_items.into_iter()
  }

  pub fn unique_items_count(&self) -> usize {
    self.unique_items.len()
  }
}

impl From<shapehash::ShapeDescriptor> for BodyDescriptor {
  fn from(mut shape_hash_descriptor: shapehash::ShapeDescriptor) -> Self {
    match shape_hash_descriptor.field_type {
      shapehash::ShapeDescriptor_PrimitiveType::OBJECT => {
        let fields = shape_hash_descriptor
          .take_fields()
          .into_iter()
          .map(|field_descriptor| {
            (
              field_descriptor.key.clone(),
              BodyDescriptor::from(field_descriptor.hash.unwrap()),
            )
          });
        BodyDescriptor::Object(ObjectDescriptor::from(fields))
      }
      shapehash::ShapeDescriptor_PrimitiveType::ARRAY => {
        let items = shape_hash_descriptor
          .take_items()
          .into_iter()
          .map(|item_descriptor| BodyDescriptor::from(item_descriptor));

        BodyDescriptor::Array(ItemsDescriptor::from(items))
      }
      shapehash::ShapeDescriptor_PrimitiveType::BOOLEAN => BodyDescriptor::Boolean,
      shapehash::ShapeDescriptor_PrimitiveType::NULL => BodyDescriptor::Null,
      shapehash::ShapeDescriptor_PrimitiveType::NUMBER => BodyDescriptor::Number,
      shapehash::ShapeDescriptor_PrimitiveType::STRING => BodyDescriptor::String,
    }
  }
}

impl From<JsonValue> for BodyDescriptor {
  fn from(json_value: JsonValue) -> Self {
    match json_value {
      JsonValue::Object(json_fields) => {
        let fields = json_fields
          .into_iter()
          .map(|(key, value)| (key, BodyDescriptor::from(value)));

        BodyDescriptor::Object(ObjectDescriptor::from(fields))
      }
      JsonValue::Array(json_items) => {
        let items = json_items
          .into_iter()
          .map(|item_descriptor| BodyDescriptor::from(item_descriptor));

        BodyDescriptor::Array(ItemsDescriptor::from(items))
      }
      JsonValue::Bool(_) => BodyDescriptor::Boolean,
      JsonValue::Null => BodyDescriptor::Null,
      JsonValue::Number(_) => BodyDescriptor::Number,
      JsonValue::String(_) => BodyDescriptor::String,
    }
  }
}

impl From<String> for BodyDescriptor {
  fn from(string: String) -> Self {
    BodyDescriptor::String
  }
}

impl From<&String> for BodyDescriptor {
  fn from(str: &String) -> Self {
    BodyDescriptor::String
  }
}

#[derive(Debug, Default)]
pub struct ParsedQueryString {
  entries: Vec<(String, String)>,
}

impl ParsedQueryString {
  pub fn from_str(query_string: &str) -> Result<Self, serde_urlencoded::de::Error> {
    let entries = serde_urlencoded::from_str(query_string)?;
    Ok(Self { entries })
  }
}

impl From<ParsedQueryString> for BodyDescriptor {
  fn from(parsed_qs: ParsedQueryString) -> Self {
    let mut values_by_key = BTreeMap::new();

    for (key, value) in parsed_qs.entries {
      let entry = values_by_key.entry(key).or_insert_with(|| vec![]);
      entry.push(value);
    }

    let fields = values_by_key.into_iter().map(|(key, values)| {
      let value_descriptor = if values.len() == 1 {
        BodyDescriptor::String
      } else {
        let item_descriptor = ItemsDescriptor::from(BodyDescriptor::String); // we only ever expect strings with this parser
        BodyDescriptor::Array(item_descriptor)
      };

      (key, value_descriptor)
    });

    BodyDescriptor::Object(ObjectDescriptor::from(fields))
  }
}

#[cfg(test)]
mod test {
  use super::*;
  use insta::assert_debug_snapshot;

  #[test]
  fn query_string_can_be_parsed_to_body_descriptor() {
    let parsed = ParsedQueryString::from_str("foo=bar&csv=1,2,51&list=twelve&list=fourteen")
      .expect("should be able to parse a query string");

    assert_debug_snapshot!(
      "query_string_can_be_parsed_to_body_descriptor__parsed",
      &parsed
    );

    let body_descriptor = BodyDescriptor::from(parsed);

    assert_debug_snapshot!(
      "query_string_can_be_parsed_to_body_descriptor__body_descriptor",
      body_descriptor
    );
  }
}
