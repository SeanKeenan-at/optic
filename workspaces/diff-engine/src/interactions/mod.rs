use crate::events::http_interaction::{Body, HttpInteraction};
use crate::learn_shape::{trail_values_for_body, TrailValueMap};
use crate::projections::{EndpointProjection, SpecProjection};
use crate::protos::shapehash::ShapeDescriptor;
use crate::queries::endpoint::EndpointQueries;
use crate::shapes::diff as diff_shape;
use crate::state::body::BodyDescriptor;

pub mod result;
mod traverser;
mod visitors;

use result::InteractionTrail;
pub use result::{BodyAnalysisResult, InteractionDiffResult};
use visitors::{InteractionVisitors, PathVisitor};

pub fn diff(
  spec_projection: &SpecProjection,
  http_interaction: HttpInteraction,
) -> Vec<InteractionDiffResult> {
  let endpoint_projection = spec_projection.endpoint();
  let endpoint_queries = EndpointQueries::new(endpoint_projection);
  let interaction_traverser = traverser::Traverser::new(&endpoint_queries);
  let mut diff_visitors = visitors::diff::DiffVisitors::new();

  interaction_traverser.traverse(&http_interaction, &mut diff_visitors);

  let results = diff_visitors.take_results().unwrap();

  results
    .into_iter()
    .flat_map(move |result| match result {
      InteractionDiffResult::MatchedRequestBodyContentType(result) => {
        // eprintln!("shape diffing for matched a request body content type");
        let body = &http_interaction.request.body.value;
        let shape_diff_results =
          diff_shape(spec_projection.shape(), body.into(), &result.root_shape_id);
        shape_diff_results
          .into_iter()
          .map(|shape_diff| {
            InteractionDiffResult::UnmatchedRequestBodyShape(
              result.clone().into_shape_diff(shape_diff),
            )
          })
          .collect()
      }
      InteractionDiffResult::MatchedResponseBodyContentType(result) => {
        // eprintln!(
        //   "interaction-diff: shape diffing for matched a response body content type: {:?}",
        //   &http_interaction.response.body
        // );
        let body = &http_interaction.response.body.value;
        let shape_diff_results =
          diff_shape(spec_projection.shape(), body.into(), &result.root_shape_id);
        shape_diff_results
          .into_iter()
          .map(|shape_diff| {
            InteractionDiffResult::UnmatchedResponseBodyShape(
              result.clone().into_shape_diff(shape_diff),
            )
          })
          .collect()
      }
      _ => vec![result],
    })
    .collect()
}

pub fn analyze_undocumented_bodies(
  spec_projection: &SpecProjection,
  interaction: HttpInteraction,
) -> impl Iterator<Item = BodyAnalysisResult> {
  let endpoint_projection = spec_projection.endpoint();
  let endpoint_queries = EndpointQueries::new(endpoint_projection);
  let interaction_traverser = traverser::Traverser::new(&endpoint_queries);
  let mut diff_visitors = visitors::diff::DiffVisitors::new();

  interaction_traverser.traverse(&interaction, &mut diff_visitors);

  let results = diff_visitors.take_results().unwrap();

  results.into_iter().filter_map(move |result| match result {
    InteractionDiffResult::UnmatchedRequestBodyContentType(diff) => {
      let body = &interaction.request.body;
      let trail_values = trail_values_for_body(&body.value);
      let interaction_trail = diff.interaction_trail.clone();
      let path_id = diff
        .requests_trail
        .get_path_id()
        .expect("UnmatchedRequestBodyContentType implies request to have a known path")
        .clone();

      Some(BodyAnalysisResult {
        path_id,
        interaction_trail,
        trail_values,
      })
    }
    InteractionDiffResult::UnmatchedResponseBodyContentType(diff) => {
      let body = &interaction.response.body;
      let trail_values = trail_values_for_body(&body.value);
      let interaction_trail = diff.interaction_trail.clone();
      let path_id = diff
        .requests_trail
        .get_path_id()
        .expect("UnmatchedResponseBodyContentType implies request to have a known path")
        .clone();

      Some(BodyAnalysisResult {
        path_id,
        interaction_trail,
        trail_values,
      })
    }
    _ => None,
  })
}
