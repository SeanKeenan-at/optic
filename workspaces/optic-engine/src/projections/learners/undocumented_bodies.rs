use cqrs_core::{Aggregate, AggregateEvent, Event};
use log;
use serde::Serialize;
use std::collections::HashMap;

use crate::commands::{EndpointCommand, SpecCommand};
use crate::interactions::{BodyAnalysisLocation, BodyAnalysisResult};
use crate::learn_shape::TrailObservationsResult;
use crate::state::SpecIdGenerator;
use crate::JsonTrail;

#[derive(Default, Debug)]
pub struct LearnedUndocumentedBodiesProjection {
  request_response_observations: HashMap<BodyAnalysisLocation, TrailObservationsResult>,
  query_params_observations: HashMap<BodyAnalysisLocation, TrailObservationsResult>,
}

impl LearnedUndocumentedBodiesProjection {
  fn with_body_analysis_result(&mut self, analysis: BodyAnalysisResult) {
    let collection = match analysis.body_location {
      BodyAnalysisLocation::UnmatchedRequest { .. }
      | BodyAnalysisLocation::UnmatchedResponse { .. } => &mut self.request_response_observations,
      BodyAnalysisLocation::UnmatchedRequestQueryParameters { .. } => {
        &mut self.query_params_observations
      }
      BodyAnalysisLocation::MatchedRequest { .. }
      | BodyAnalysisLocation::MatchedResponse { .. } => {
        unreachable!("analyzing undocumented bodies should only yield unmatched analysiss")
      }
    };

    let existing_observations = collection
      .entry(analysis.body_location)
      .or_insert_with(|| TrailObservationsResult::default());

    existing_observations.union(analysis.trail_observations.normalized());
  }

  pub fn into_endpoint_bodies(
    mut self,
    id_generator: &mut impl SpecIdGenerator,
  ) -> impl Iterator<Item = EndpointBodies> {
    let mut endpoints_by_endpoint = HashMap::new();
    for (body_location, observations) in self.request_response_observations {
      let (root_shape_id, body_commands) =
        observations.into_commands(id_generator, &JsonTrail::empty());
      let mut endpoint_body = EndpointBody::new(&body_location, root_shape_id, body_commands);

      if let BodyAnalysisLocation::UnmatchedRequest {
        ref path_id,
        ref method,
        ref content_type,
      } = body_location
      {
        let expected_query_params_location =
          BodyAnalysisLocation::UnmatchedRequestQueryParameters {
            path_id: path_id.clone(),
            method: method.clone(),
            content_type: content_type.clone(),
          };
        let query_params_analysis = self
          .query_params_observations
          .remove(&expected_query_params_location);

        if let Some(query_observations) = query_params_analysis {
          let (query_parameters_shape_id, query_parameters_commands) =
            query_observations.into_commands(id_generator, &JsonTrail::empty());

          if let Some(query_parameters_shape_id) = query_parameters_shape_id {
            endpoint_body.with_query_params(query_parameters_shape_id, query_parameters_commands)
          }
        }
      }

      endpoint_body.append_endpoint_commands(id_generator);

      let (path_id, method) = match body_location {
        BodyAnalysisLocation::UnmatchedRequest {
          path_id, method, ..
        } => (path_id, method),
        BodyAnalysisLocation::UnmatchedResponse {
          path_id, method, ..
        } => (path_id, method),
        _ => unreachable!("analyzing undocumented bodies should only yield unmatched results"),
      };

      let endpoint_bodies = endpoints_by_endpoint
        .entry((path_id, method))
        .or_insert_with_key(|(path_id, method)| {
          EndpointBodies::new(path_id.clone(), method.clone())
        });

      endpoint_bodies.push(endpoint_body);
    }

    endpoints_by_endpoint.into_iter().map(|(k, v)| v)
  }
}

impl Aggregate for LearnedUndocumentedBodiesProjection {
  fn aggregate_type() -> &'static str {
    "learned_undocument_bodies"
  }
}

impl Event for BodyAnalysisResult {
  fn event_type(&self) -> &'static str {
    "body_analysis_result"
  }
}

impl AggregateEvent<LearnedUndocumentedBodiesProjection> for BodyAnalysisResult {
  fn apply_to(self, aggregate: &mut LearnedUndocumentedBodiesProjection) {
    aggregate.with_body_analysis_result(self)
  }
}

// Output structs
// --------------

#[derive(Default, Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct EndpointBodies {
  path_id: String,
  method: String,
  requests: Vec<EndpointRequestBody>,
  responses: Vec<EndpointResponseBody>,
}

impl EndpointBodies {
  pub fn new(path_id: String, method: String) -> Self {
    Self {
      path_id,
      method,
      requests: vec![],
      responses: vec![],
    }
  }

  pub fn push(&mut self, endpoint: EndpointBody) {
    match endpoint {
      EndpointBody::Request(endpoint_request) => {
        self.requests.push(endpoint_request);
      }
      EndpointBody::Response(endpoint_response) => {
        self.responses.push(endpoint_response);
      }
    }
  }
}

#[derive(Debug)]
pub enum EndpointBody {
  Request(EndpointRequestBody),
  Response(EndpointResponseBody),
}

#[derive(Default, Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct EndpointRequestBody {
  commands: Vec<SpecCommand>,

  #[serde(skip)]
  path_id: String,
  #[serde(skip)]
  method: String,
  #[serde(skip)]
  query_parameters_shape_id: Option<String>,

  #[serde(flatten)]
  body_descriptor: Option<EndpointBodyDescriptor>,
}

#[derive(Default, Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct EndpointResponseBody {
  commands: Vec<SpecCommand>,
  status_code: u16,

  #[serde(skip)]
  path_id: String,
  #[serde(skip)]
  method: String,

  #[serde(flatten)]
  body_descriptor: Option<EndpointBodyDescriptor>,
}

#[derive(Default, Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct EndpointBodyDescriptor {
  content_type: String,
  root_shape_id: String,
}

impl EndpointBody {
  fn new(
    body_location: &BodyAnalysisLocation,
    root_shape_id: Option<String>,
    body_commands: impl IntoIterator<Item = SpecCommand>,
  ) -> Self {
    let body_descriptor = match root_shape_id {
      Some(root_shape_id) => Some(EndpointBodyDescriptor {
        content_type: body_location
          .content_type()
          .expect("root shape id implies a content type to be present")
          .clone(),
        root_shape_id,
      }),
      None => None,
    };

    match body_location {
      BodyAnalysisLocation::UnmatchedRequest {
        path_id, method, ..
      } => EndpointBody::Request(EndpointRequestBody {
        body_descriptor,
        path_id: path_id.clone(),
        method: method.clone(),
        commands: body_commands.into_iter().collect(),
        query_parameters_shape_id: None,
      }),
      BodyAnalysisLocation::UnmatchedResponse {
        status_code,
        path_id,
        method,
        ..
      } => EndpointBody::Response(EndpointResponseBody {
        body_descriptor,
        path_id: path_id.clone(),
        method: method.clone(),
        commands: body_commands.into_iter().collect(),
        status_code: *status_code,
      }),
      _ => panic!("EndpointBody should only be created for unmatched responses and requests"),
    }
  }

  fn with_query_params(
    &mut self,
    root_shape_id: String,
    body_commands: impl IntoIterator<Item = SpecCommand>,
  ) {
    match self {
      EndpointBody::Request(request_body) => {
        request_body.query_parameters_shape_id = Some(root_shape_id);
        request_body.commands.extend(body_commands);
      }
      EndpointBody::Response(_) => {
        panic!("query parameters can only be added to the Request variant of an EndpointBody")
      }
    }
  }

  fn append_endpoint_commands(&mut self, ids: &mut impl SpecIdGenerator) {
    match self {
      EndpointBody::Request(request_body) => {
        let request_id = ids.request();
        request_body
          .commands
          .push(SpecCommand::from(EndpointCommand::add_request(
            request_id.clone(),
            request_body.path_id.clone(),
            request_body.method.clone(),
          )));

        if let Some(body_descriptor) = &request_body.body_descriptor {
          request_body
            .commands
            .push(SpecCommand::from(EndpointCommand::set_request_body_shape(
              request_id.clone(),
              body_descriptor.root_shape_id.clone(),
              body_descriptor.content_type.clone(),
              false,
            )));
        }

        if let Some(query_parameters_shape_id) = &request_body.query_parameters_shape_id {
          request_body.commands.push(SpecCommand::from(
            EndpointCommand::set_request_query_parameters_shape(
              request_id.clone(),
              query_parameters_shape_id.clone(),
              false,
            ),
          ))
        }

        // TODO: append query parameter commands
      }
      EndpointBody::Response(response_body) => {
        let response_id = ids.response();
        response_body.commands.push(SpecCommand::from(
          EndpointCommand::add_response_by_path_and_method(
            response_id.clone(),
            response_body.path_id.clone(),
            response_body.method.clone(),
            response_body.status_code.clone(),
          ),
        ));

        if let Some(body_descriptor) = &response_body.body_descriptor {
          response_body
            .commands
            .push(SpecCommand::from(EndpointCommand::set_response_body_shape(
              response_id,
              body_descriptor.root_shape_id.clone(),
              body_descriptor.content_type.clone(),
              false,
            )));
        }
      }
    };
  }
}

#[cfg(test)]
mod test {
  use super::*;
  use crate::events::SpecEvent;
  use crate::interactions::InteractionDiffResult;
  use crate::learn_shape::observe_body_trails;
  use crate::projections::SpecProjection;
  use crate::state::body::BodyDescriptor;
  use insta::assert_debug_snapshot;
  use serde_json::json;

  #[test]
  fn undocumented_bodies_can_aggregate_analysis_results_with_array_items() {
    let body = BodyDescriptor::from(json!({
      "items": [132, "string-array-item"],
      "other-field": true
    }));

    let analysis_result = BodyAnalysisResult {
      body_location: BodyAnalysisLocation::UnmatchedResponse {
        path_id: String::from("test-path-1"),
        method: String::from("GET"),
        content_type: Some(String::from("application/json")),
        status_code: 200,
      },
      trail_observations: observe_body_trails(body),
    };

    let mut projection = LearnedUndocumentedBodiesProjection::default();

    projection.apply(analysis_result.clone());

    // dbg!(&projection);
    let aggregated_result = projection
      .request_response_observations
      .get(&analysis_result.body_location)
      .unwrap();

    let array_trail = JsonTrail::empty().with_object_key(String::from("items"));

    let items_values = aggregated_result
      .values_by_trail
      .get(&array_trail.with_array_item(0))
      .expect("should have learned values for items array");
    assert!(items_values.was_number && items_values.was_string);
    assert!(
      aggregated_result
        .values_by_trail
        .get(&array_trail.with_array_item(1))
        .is_none(),
      "undocumented endpoints projection aggregates normalized body results",
    );

    let mut aggregated_trails: Vec<_> = aggregated_result.trails().collect();
    aggregated_trails.sort();

    assert_debug_snapshot!(
      "undocumented_bodies_can_aggregate_analysis_results_with_array_items__aggregated_trails",
      aggregated_trails
    );
  }

  #[test]
  fn undocumented_bodies_generates_commands_for_request_query_parameters() {
    let test_path = "root";
    let test_method = "GET";
    let query_params_body = BodyDescriptor::from(json!({
      "search": "a-search-query",
      "page": "3"
    }));

    let analysis_results = vec![
      BodyAnalysisResult {
        body_location: BodyAnalysisLocation::UnmatchedRequest {
          content_type: None,
          path_id: String::from(test_path),
          method: String::from(test_method),
        },
        trail_observations: observe_body_trails(None),
      },
      BodyAnalysisResult {
        // query params matching the preceding request
        body_location: BodyAnalysisLocation::UnmatchedRequestQueryParameters {
          path_id: String::from(test_path),
          method: String::from(test_method),
          content_type: None,
        },
        trail_observations: observe_body_trails(query_params_body.clone()),
      },
      BodyAnalysisResult {
        // query params for a different request
        body_location: BodyAnalysisLocation::UnmatchedRequestQueryParameters {
          path_id: String::from(test_path),
          method: String::from(test_method),
          content_type: Some(String::from("application/json")),
        },
        trail_observations: observe_body_trails(query_params_body),
      },
    ];

    let mut test_id_generator = TestIdGenerator::default();
    let mut projection = LearnedUndocumentedBodiesProjection::default();

    for result in analysis_results {
      projection.apply(result);
    }

    let mut endpoint_bodies = projection
      .into_endpoint_bodies(&mut test_id_generator)
      .collect::<Vec<_>>();
    assert_eq!(endpoint_bodies.len(), 1);

    let mut endpoint_body = endpoint_bodies.remove(0);
    assert_eq!(endpoint_body.requests.len(), 1);

    let request_commands = endpoint_body.requests.remove(0).commands;

    assert_debug_snapshot!(
      "undocumented_bodies_generates_commands_for_request_query_parameters__request_commands",
      &request_commands
    );

    let base_spec = SpecProjection::default();
    assert_valid_commands(base_spec, request_commands);
  }

  #[test]
  fn undocumented_bodies_can_generate_commands_for_request_with_empty_query_parameters() {
    let test_path = "root";
    let test_method = "GET";

    let analysis_results = vec![
      BodyAnalysisResult {
        body_location: BodyAnalysisLocation::UnmatchedRequest {
          content_type: None,
          path_id: String::from(test_path),
          method: String::from(test_method),
        },
        trail_observations: observe_body_trails(None),
      },
      BodyAnalysisResult {
        body_location: BodyAnalysisLocation::UnmatchedRequestQueryParameters {
          path_id: String::from(test_path),
          method: String::from(test_method),
          content_type: None,
        },
        trail_observations: observe_body_trails(None),
      },
    ];

    let mut test_id_generator = TestIdGenerator::default();
    let mut projection = LearnedUndocumentedBodiesProjection::default();

    for result in analysis_results {
      projection.apply(result);
    }

    let mut endpoint_bodies = projection
      .into_endpoint_bodies(&mut test_id_generator)
      .collect::<Vec<_>>();
    assert_eq!(endpoint_bodies.len(), 1);

    let mut endpoint_body = endpoint_bodies.remove(0);
    assert_eq!(endpoint_body.requests.len(), 1);

    let request_commands = endpoint_body.requests.remove(0).commands;

    assert_debug_snapshot!(
      "undocumented_bodies_can_generate_commands_for_request_with_empty_query_parameters__request_commands",
      &request_commands
    );

    let base_spec = SpecProjection::default();
    assert_valid_commands(base_spec, request_commands);
  }

  #[test]
  fn undocumented_bodies_generates_commands_for_responses_without_bodies() {
    let analysis_result = BodyAnalysisResult {
      body_location: BodyAnalysisLocation::UnmatchedResponse {
        content_type: None,
        path_id: String::from("path-1"),
        method: String::from("DELETE"),
        status_code: 204,
      },
      trail_observations: observe_body_trails(None),
    };

    let mut test_id_generator = TestIdGenerator::default();
    let mut projection = LearnedUndocumentedBodiesProjection::default();

    projection.apply(analysis_result);

    let endpoint_bodies = projection
      .into_endpoint_bodies(&mut test_id_generator)
      .collect::<Vec<_>>();

    assert_debug_snapshot!(
      "undocumented_bodies_generates_commands_for_responses_without_bodies__endpoint_bodies",
      endpoint_bodies
    );
  }

  #[derive(Debug, Default)]
  struct TestIdGenerator {
    counter: usize,
  }

  impl SpecIdGenerator for TestIdGenerator {
    fn generate_id(&mut self, _prefix: &str) -> String {
      let id = format!("test-id-{}", self.counter);
      self.counter += 1;
      id
    }
  }

  fn assert_valid_commands(
    mut spec_projection: SpecProjection,
    commands: impl IntoIterator<Item = SpecCommand>,
  ) -> SpecProjection {
    // let mut spec_projection = SpecProjection::default();
    for command in commands {
      let events = spec_projection
        .execute(command)
        .expect("generated commands must be valid");

      for event in events {
        spec_projection.apply(event)
      }
    }

    spec_projection
  }
}
