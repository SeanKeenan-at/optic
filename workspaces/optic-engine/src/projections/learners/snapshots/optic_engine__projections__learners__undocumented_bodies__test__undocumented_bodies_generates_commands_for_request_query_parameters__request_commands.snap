---
source: workspaces/optic-engine/src/projections/learners/undocumented_bodies.rs
expression: "&request_commands"
---
[
    ShapeCommand(
        AddShape(
            AddShape {
                shape_id: "test-id-0",
                base_shape_id: "$string",
                name: "",
            },
        ),
    ),
    ShapeCommand(
        AddShape(
            AddShape {
                shape_id: "test-id-1",
                base_shape_id: "$string",
                name: "",
            },
        ),
    ),
    ShapeCommand(
        AddShape(
            AddShape {
                shape_id: "test-id-4",
                base_shape_id: "$object",
                name: "",
            },
        ),
    ),
    ShapeCommand(
        AddField(
            AddField {
                field_id: "test-id-2",
                shape_id: "test-id-4",
                name: "page",
                shape_descriptor: FieldShapeFromShape(
                    FieldShapeFromShape {
                        field_id: "test-id-2",
                        shape_id: "test-id-1",
                    },
                ),
            },
        ),
    ),
    ShapeCommand(
        AddField(
            AddField {
                field_id: "test-id-3",
                shape_id: "test-id-4",
                name: "search",
                shape_descriptor: FieldShapeFromShape(
                    FieldShapeFromShape {
                        field_id: "test-id-3",
                        shape_id: "test-id-0",
                    },
                ),
            },
        ),
    ),
    EndpointCommand(
        AddRequest(
            AddRequest {
                request_id: "test-id-5",
                path_id: "root",
                http_method: "GET",
            },
        ),
    ),
    EndpointCommand(
        SetRequestQueryParametersShape(
            SetRequestQueryParametersShape {
                request_id: "test-id-5",
                shape_descriptor: QueryParametersShapeDescriptor {
                    shape_id: "test-id-4",
                    is_removed: false,
                },
            },
        ),
    ),
]
